Insert image in format consistent with pandoc

gadenbuie opened the pull request:
First, thank you @lcolladotor, the Insert Image addin is 💯 👍 

I noticed that the addin uses `shiny::img()` to create the HTML img tag when width or height are set in the addin. When `height` or `width` are not set, though, the addin uses standard markdown format of `![alt](src)`, which when processed by pandoc is converted to

```html
<div class="figure">
<img src="{{src}}" alt="{{alt}}"/>
<p class="caption">{{alt}}</p>
</div>
```

Using `shiny::img()` then has the side effect of removing the caption and results in an image with slightly different formatting.

This PR adds the function `figure_code()` to the Insert Image addin that adds width or height into the `<img>` tag inside the figure `div` to be consistent with pandoc's `![]()` processing.

For example, inserting an image and setting the width or height param using the gadget

![screenshot 2018-03-15 23 29 09](https://user-images.githubusercontent.com/5420529/37502946-0712c940-28ac-11e8-9b04-eb40fc6e7a5f.png)

now inserts this code in the `.Rmd` (`.md`) and the `.html` file.

```html
<div class="figure">
<img src="/post/2018-03-15-test-insert-image_files/hadley_typing2.gif" alt="Hadley Typing (width set)" width="50%"/>
<p class="caption">Hadley Typing (width set)</p>
</div>
```

Not setting width results in this in the `.Rmd` file

```md
![Hadley Typing (w/h not set)](/post/2018-03-15-test-insert-image_files/hadley_typing.gif)
```

which is converted by pandoc to this in the `.html`

```html
<div class="figure">
<img src="/post/2018-03-15-test-insert-image_files/hadley_typing.gif" alt="Hadley Typing (w/h not set)" />
<p class="caption">Hadley Typing (w/h not set)</p>
</div>
```

For reference, this was the previous output of the insert image addin

```html
<img src="/post/2018-03-15-test-insert-image_files/hadley_typing2.gif" alt="Hadley Typing (shiny::img)" width="33%"/>
```

And they all look good in the final post

![screenshot 2018-03-15 23 57 06](https://user-images.githubusercontent.com/5420529/37503102-d4ea6864-28ac-11e8-8897-19b705d5a416.png)

gadenbuie commented:
Oops! I forgot that Pandoc now supports those attributes. That’s way better, I also wasn’t super happy with inserting 5 lines of HTML into the post. And the use of `shiny::img()` makes a lot of sense regarding escaping html. 

I’ll update the PR this afternoon.

lcolladotor commented:
Thanks for the shoutout!

Cool, I didn't know that `Pandoc actually supports width/height attributes in Markdown`.

gadenbuie commented:
I've updated the PR in d68b12d to revert to previous behavior (`![alt](src)` if no width/height or `shing::img()` if width height) unless the selected post is an `.Rmd` file, in which case `![alt](src){width=W height=H}` is used.

Because we're now using the selected file extension, I added [a check at the top](https://github.com/gadenbuie/blogdown/blob/d68b12d19e721db85684e909a2d2bca88ae4b94c/inst/scripts/insert_image.R#L8-L17) that throws an error if the selected file is not an `.Rmd`, `.md`, or `.Rmarkdown` file, in the style of the previous check that a file is actually selected.

Commit 34c2eaf goes a little further and runs a series of checks to make sure that the width/height attributes are consistent with [Pandoc's style guide](http://pandoc.org/MANUAL.html#extension-link_attributes). Essentially, I strip spaces, check that the value starts with a number, and check that the unit (if provided) is valid. Because a bad width/height attribute won't break the site rendering -- but it will probably look weird -- bad formatting leads to a warning only with instructions to review the inserted image source.

The warnings read as follows

```r
> check_pandoc_format("\\textwidth")
Warning message:
Please review the attributes of the inserted image source.
Width/height attributes must start with a number, followed by unit without spaces and unit must be one of `px`, `cm`, `mm`, `in`, `inch`, or `%` 

> check_pandoc_format("200em")
Warning message:
Please review the attributes of the inserted image source.
Width/height unit must be one of `px`, `cm`, `mm`, `in`, `inch`, or `%` 

> check_pandoc_format("px")
Warning message:
Please review the attributes of the inserted image source.
Width/height attributes must start with a number, followed by unit without spaces 
```

If this goes a little too far, feel free to revert 34c2eaf.

yihui commented on inst/scripts/insert_image.R:
```diff
@@ -5,6 +5,16 @@ local({
+    "rmd" = "rmd",
+    "md" = "md",
+    "rmarkdown" = "md",
+    stop(
```
I hope that we only check for `.rmd` and use the `![](){}` syntax. Other extensions are not necessarily problematic in Hugo (e.g. Hugo supports AsciiDoc).

gadenbuie replied:
So the check I added does limit the "Insert Image" addin to files with extension `rmd`, `md` or `rmarkdown`. The version in the master branch currently assumes that markdown syntax will be valid unless width/height need to be set, in which case it uses HTML.

Based on your comment, what I can do is check the file extension and make sure it's one of [the Hugo-supported source files](https://github.com/gohugoio/hugo/blob/77c60a3440806067109347d04eb5368b65ea0fe8/helpers/general.go#L67-L78). 

Then we can use `![]()` if we're in a markdown-capable source file (without width/height set), or `![](){}` if we're in an `.Rmd` file (with w/h set). Otherwise, we can use `shiny::img()` for standard HTML for any other extensions.

I'm assuming that AsciiDoc, RST pass through HTML, as the image syntax in those formats is different from markdown.

If the file is not a potential blog source file (say an R script), should we exit, warn or add some other kind of interaction (like a pop-up modal confirming the user wants to add the image to the file)?

yihui replied:
I mean we don't need to be strict here. If the file is .Rmd or the width/height are not set, we use the Markdown syntax. For all other cases, use HTML. If the user tries to insert an image to an R script, I'd say just let it go (without checking anything on our side, and we assume he/she knows what he/she is doing).

yihui commented on inst/scripts/insert_image.R:
```diff
@@ -69,9 +79,43 @@ local({
+            paste0('![', alt, '](', s, ')')
+          } else {
+            if (ctx_ext == "rmd") {
+              check_pandoc_format = function(x) {
```
You are basically reinventing `htmltools::validateCssUnit()` in this function :) You can probably use `try(htmltools::validateCssUnit())` and turn the possible error to a warning.

gadenbuie replied:
Yup, that's pretty much exactly what I did 🤣! I'll update to use `htmltools::validateCssUnit()` rather than reinventing the wheel.

yihui commented on inst/scripts/insert_image.R:
```diff
@@ -69,9 +79,28 @@ local({
+              if (x == '') return(x)
+              tryCatch(
+                htmltools::validateCssUnit(x),
+                error = function(e) warning(e$message, call. = FALSE)
```
need to return `x` in the error handler:

```r
error = function(e) {
  warning(e$message, call. = FALSE)
  x
}
```

gadenbuie commented:
Ok, I think we're on the same page now. The addin just checks to see if current file is an `.Rmd`, in which case pandoc syntax is used, otherwise it remains as before (as described in https://github.com/rstudio/blogdown/pull/279#discussion_r175874113).

You'll squash this into a single commit right? 😉 I can clean up into a better PR if desired.

yihui commented:
Yes, I'll squash and merge the PR. Thanks!

lcolladotor commented:
Cool! I'm looking forward to using this new version =)
